# Week9 Ganesh Kale

http://thinkstats2.com

Copyright 2016 Allen B. Downey

MIT License: https://opensource.org/licenses/MIT


In [1]:
from __future__ import print_function, division

%matplotlib inline

import numpy as np
import pandas as pd

import random

import thinkstats2
import thinkplot

## Exercises 11.1

**Exercise:** Suppose one of your co-workers is expecting a baby and you are participating in an office pool to predict the date of birth. Assuming that bets are placed during the 30th week of pregnancy, what variables could you use to make the best prediction? You should limit yourself to variables that are known before the birth, and likely to be available to the people in the pool.

In [101]:
import first
live, firsts, others = first.MakeFrames()
live = live[live.prglngth>30]

The following are the only variables I found that have a statistically significant effect on pregnancy length.

In [102]:
# added ager columns as independent variable to test the age is statistically significant or not
# but based on the p-value it seems not significant

In [103]:
import statsmodels.formula.api as smf
model = smf.ols('prglngth ~ birthord==1 + race==2 + nbrnaliv>1 + ager', data=live)
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               prglngth   R-squared:                       0.011
Model:                            OLS   Adj. R-squared:                  0.011
Method:                 Least Squares   F-statistic:                     25.80
Date:                Fri, 14 May 2021   Prob (F-statistic):           2.71e-21
Time:                        21:55:24   Log-Likelihood:                -18247.
No. Observations:                8884   AIC:                         3.650e+04
Df Residuals:                    8879   BIC:                         3.654e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
=========================================================================================
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
Intercept                38.6976      0.111    349.048      0.000      38.480      38.915
birthord == 1[T.True]     0.1044      0.040      2.583      0.010       0.025       0.184
race == 2[T.True]         0.1371      0.042      3.257      0.001       0.055       0.220
nbrnaliv > 1[T.True]     -1.4964      0.165     -9.095      0.000      -1.819      -1.174
ager                      0.0019      0.003      0.616      0.538      -0.004       0.008
==============================================================================
Omnibus:                     1585.512   Durbin-Watson:                   1.620
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             6158.217
Skew:                          -0.850   Prob(JB):                         0.00
Kurtosis:                       6.707   Cond. No.                         282.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [104]:
import statsmodels.formula.api as smf
model = smf.ols('prglngth ~ birthord==1 + race==2 + nbrnaliv>1', data=live)
results_mlr = model.fit()
results_mlr.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               prglngth   R-squared:                       0.011
Model:                            OLS   Adj. R-squared:                  0.011
Method:                 Least Squares   F-statistic:                     34.28
Date:                Fri, 14 May 2021   Prob (F-statistic):           5.09e-22
Time:                        21:55:25   Log-Likelihood:                -18247.
No. Observations:                8884   AIC:                         3.650e+04
Df Residuals:                    8880   BIC:                         3.653e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=========================================================================================
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
Intercept                38.7617      0.039   1006.410      0.000      38.686      38.837
birthord == 1[T.True]     0.1015      0.040      2.528      0.011       0.023       0.180
race == 2[T.True]         0.1390      0.042      3.311      0.001       0.057       0.221
nbrnaliv > 1[T.True]     -1.4944      0.164     -9.086      0.000      -1.817      -1.172
==============================================================================
Omnibus:                     1587.470   Durbin-Watson:                   1.619
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             6160.751
Skew:                          -0.852   Prob(JB):                         0.00
Kurtosis:                       6.707   Cond. No.                         10.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### Based on the above result generated by statstmodel multiple regression, the variables that would make the best predictions about data of birth are - 
1. birth order = 1 - the responsents first child
2. race - race of the responsents
3. number of live births - delivered multiple babies - twins etc
4. age - age of the respondent (ager) - **but this is not statistically significant

## Exercises 11.3

**Exercise:** If the quantity you want to predict is a count, you can use Poisson regression, which is implemented in StatsModels with a function called `poisson`. It works the same way as `ols` and `logit`. As an exercise, let’s use it to predict how many children a woman has born; in the NSFG dataset, this variable is called `numbabes`.

Suppose you meet a woman who is 35 years old, black, and a college graduate whose annual household income exceeds $75,000. How many children would you predict she has born?

In [105]:
live, firsts, others = first.MakeFrames()
resp = nsfg.ReadFemResp()
resp.index = resp.caseid
join = live.join(resp, on='caseid', rsuffix='_r')

In [106]:

model = smf.poisson('numbabes ~ ager==35 + race==1 + educat==16 + totincr', data=join)
results_p = model.fit()
results_p.summary()

Optimization terminated successfully.
         Current function value: 1.721278
         Iterations 4


<class 'statsmodels.iolib.summary.Summary'>
"""
                          Poisson Regression Results                          
==============================================================================
Dep. Variable:               numbabes   No. Observations:                 9148
Model:                        Poisson   Df Residuals:                     9143
Method:                           MLE   Df Model:                            4
Date:                Fri, 14 May 2021   Pseudo R-squ.:                 0.01159
Time:                        21:55:50   Log-Likelihood:                -15746.
converged:                       True   LL-Null:                       -15931.
Covariance Type:            nonrobust   LLR p-value:                 1.219e-78
========================================================================================
                           coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------------
Intercept                1.1480      0.016     71.961      0.000       1.117       1.179
ager == 35[T.True]      -0.0704      0.032     -2.226      0.026      -0.132      -0.008
race == 1[T.True]        0.1337      0.014      9.479      0.000       0.106       0.161
educat == 16[T.True]    -0.0589      0.024     -2.492      0.013      -0.105      -0.013
totincr                 -0.0219      0.002    -12.833      0.000      -0.025      -0.019
========================================================================================
"""

Now we can predict the number of children for a woman who is 35 years old, black, and a college
graduate whose annual household income exceeds $75,000

In [107]:
# predict the how many children she has born - 

columns = ['ager', 'race', 'totincr', 'educat']
df = pd.DataFrame([[35, 1, 14, 16]], columns=columns)
f"The woman would have {round(results_p.predict(df)[0])} childrens"

'The woman would have 2 childrens'

## Exercises 11.3

**Exercise:** If the quantity you want to predict is categorical, you can use multinomial logistic regression, which is implemented in StatsModels with a function called `mnlogit`. As an exercise, let’s use it to guess whether a woman is married, cohabitating, widowed, divorced, separated, or never married; in the NSFG dataset, marital status is encoded in a variable called `rmarital`.

Suppose you meet a woman who is 25 years old, white, and a high school graduate whose annual household income is about $45,000. What is the probability that she is married, cohabitating, etc?

In [112]:
model = smf.mnlogit('rmarital ~ ager + C(race) + totincr + educat', data=join)
results_mnl = model.fit()
results_mnl.summary()

Optimization terminated successfully.
         Current function value: 1.095371
         Iterations 8


<class 'statsmodels.iolib.summary.Summary'>
"""
                          MNLogit Regression Results                          
==============================================================================
Dep. Variable:               rmarital   No. Observations:                 9148
Model:                        MNLogit   Df Residuals:                     9118
Method:                           MLE   Df Model:                           25
Date:                Fri, 14 May 2021   Pseudo R-squ.:                  0.1636
Time:                        21:58:30   Log-Likelihood:                -10020.
converged:                       True   LL-Null:                       -11981.
Covariance Type:            nonrobust   LLR p-value:                     0.000
================================================================================
  rmarital=2       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        4.4558      0.273     16.296      0.000       3.920       4.992
C(race)[T.2]    -0.9241      0.086    -10.692      0.000      -1.093      -0.755
C(race)[T.3]    -0.6456      0.133     -4.863      0.000      -0.906      -0.385
ager            -0.0574      0.006    -10.031      0.000      -0.069      -0.046
totincr         -0.1293      0.011    -11.433      0.000      -0.151      -0.107
educat          -0.2033      0.018    -11.110      0.000      -0.239      -0.167
--------------------------------------------------------------------------------
  rmarital=3       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -4.3815      0.903     -4.850      0.000      -6.152      -2.611
C(race)[T.2]    -0.3953      0.235     -1.684      0.092      -0.855       0.065
C(race)[T.3]     0.0418      0.334      0.125      0.900      -0.613       0.697
ager             0.1271      0.019      6.782      0.000       0.090       0.164
totincr         -0.3206      0.032    -10.134      0.000      -0.383      -0.259
educat          -0.1112      0.048     -2.318      0.020      -0.205      -0.017
--------------------------------------------------------------------------------
  rmarital=4       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -2.6453      0.297     -8.913      0.000      -3.227      -2.064
C(race)[T.2]    -0.3278      0.091     -3.612      0.000      -0.506      -0.150
C(race)[T.3]    -0.8179      0.169     -4.826      0.000      -1.150      -0.486
ager             0.0649      0.006     10.104      0.000       0.052       0.078
totincr         -0.2283      0.011    -20.020      0.000      -0.251      -0.206
educat           0.0668      0.016      4.059      0.000       0.035       0.099
--------------------------------------------------------------------------------
  rmarital=5       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        0.7899      0.340      2.326      0.020       0.124       1.455
C(race)[T.2]    -1.0567      0.101    -10.443      0.000      -1.255      -0.858
C(race)[T.3]    -0.6057      0.154     -3.932      0.000      -0.908      -0.304
ager             0.0073      0.007      1.030      0.303      -0.007       0.021
totincr         -0.2893      0.014    -20.309      0.000      -0.317      -0.261
educat          -0.0144      0.021     -0.685      0.493      -0.056       0.027
--------------------------------------------------------------------------------
  rmarital=6       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        6.2571      0.270     23.194      0.000       5.728       6.786
C

Make a prediction for a woman who is 25 years old, white, and a high
school graduate whose annual household income is about $45,000.

In [113]:
columns = ['ager', 'race', 'totincr', 'educat']
df_mnl = pd.DataFrame([[25, 2, 11, 12]], columns=columns)
results_mnl.predict(df_mnl)

0         1         2         3         4         5
0  0.744248  0.127276  0.001165  0.034928  0.023891  0.068492

0 - Currently married
1 - Not married but living with opp sex partner
2 - widowed
3 - divorced
4 - separated for reasons of marital discord
5 - Never been married

Based on above results we can say that - there are 74% chances that woman is married or 13% chances that living with opp sex partner or very minimal chances of widowed or divorced or never married, separated etc


END